### ISSUE TITLE
`gesture_overview_enable = false` disables workspace swipe

### ISSUE THREAD

**Issue by Acumane** [dawsers/hyprscroller#96]
Would expect this to disable only the overview gesture but it disables all 4-fingered gestures instead.

**Comment by dawsers** [comment]
This is explained in the [README](https://github.com/dawsers/hyprscroller#gesture_overview_enable).

The reason is most touchpads only have three and four finger events available (aside from the standard one or two finger gestures). For scrolling I need the two axes, but overview and workspaces only need one axis each.

It could be clearer and separate the `_enable` option, but that would also make it less clear that `_fingers` should  be the same value. I don't really want to add personalized gestures, and people choices for whether to swipe up/down or to the sides. I think the default is quite reasonable. I can add `gesture_workspaces_enable`, `gestures_workspaces_fingers` and `gesture_workspaces_distance` if you think that helps you, but my fear is people will start changing `fingers` not realizing the reason why they are shared.

**Comment by dawsers** [comment]
The default Hyprland workspace swipe doesn't work with *hyprscroller*. The explanation is [here](https://github.com/dawsers/hyprscroller/issues/73#issuecomment-2480687142).

*hyprscroller* has workspace windows outside of the monitor, while Hyprland doesn't. so when you swipe to change workspace, Hyprland can show a bit of the active workspace and a bit of the next in the animation, because they don't overlap. Instead, if you tried the same with *hyprscroller*, it would show windows that are outside of the viewport, but have the same coordinates as the ones in the next workspace (coordinate space overlaps).

So basically, workspace swipe cannot work the same way as Hyprland's, because workspaces have overlapping coordinates (they are infinite). No fancy animations, sorry.

**Comment by Acumane** [comment]
Back to the original issue, it is simply the case that `gesture_overview_enable` misleads. This matters to some users (me) because I'm using [fusuma](https://github.com/iberianpig/fusuma) to register a 4-fingered vertical swipe for brightness, and I'd like workspace swipe to keep working.

**Comment by dawsers** [comment]
Yes, I can separate the two.

**Comment by dawsers** [comment]
Impemented in af08f46b8706572d3e3f2e030e6687934aa66fe3

**Comment by BinaryInk** [comment]
This is by *no means* important as it's *just* eye candy, but I forgot about workspace swiping via touchscreen and accidentally initiated it, then I remembered reading this:

> So basically, workspace swipe cannot work the same way as Hyprland's, because workspaces have overlapping coordinates (they are infinite). No fancy animations, sorry.

It *would* seem the 'fancy animations' *are* possible (as far as I can tell), as the Hyprland implementation of touchscreen workspace swipes do work with scroller enabled, animations and all...

----

Fortunately, I have fixed my XDG-Portal/OBS issue in the past couple weeks and can demonstrate:

https://github.com/user-attachments/assets/a284e39f-8c63-4a7f-b58d-223a036b79db

----

I don't know the technical details, so it could still be impossible to implement or difficult to implement--I don't know how important you'd see that being to implement if it is in fact possible... but thought I'd point it out nevertheless if it were something you'd want to put on the backburner to address.

But, again: it's eye candy, the current implementation with the touchpad looks and works *just* fine.

**Comment by dawsers** [comment]
Do you have more than one monitor enabled? I think you probably only have one in that video.

When you have two, workspace numbers alternate. So when you move from let's say 9 to 10, you are changing monitor from M1 to M2. The problem shows there, the coordinates of monitor 2 (M2) overlap the scroller workspace for M1.

**Comment by dawsers** [comment]
Two monitors next to each other in landscape mode in Hyprland have coordinates M1=[0,w1, 0,h1] and M2=[w1, w1+w2, 0, h2].

As roughly explained before, Master and Dwindle don't use a virtual workspace, Simplifying, in Hyprland physical and virtual coordinates map one to one. Their extents are those of the monitor. So when the animation happens moving from one monitor to the other, the other monitor has a disjoint set of coordinates. To simplify what's happening, they can show windows from both monitors by simply moving the viewport in an animation. That viewport will only have wndows from one workspace or the other at each pixel, because their coordinates don't overlap.

Now think hyprscroller. Your viewport in M1 has physical coordinates from [0->w, 0->h], and in M2 from [w1-> w1+w2, 0-> h2], but the workspaces have coordinates that may spawn an infinite space [-inf, inf, -inf, inf]. So when the gesture tries to animate a moving viewport and it doesn't know anything about hyprscroller, it will choose those windows with coordinates inside the animated, moving viewport. So it will show windows from both workspaces all the time, because they overlap with physical coordinates of the other monitor. For each pixel, you can have windows from both workspaces.

I will verify what you are seeing, but I doubt they have implemented virtual workspaces, so it won't probably work with setups that have more than one monitor.

**Comment by BinaryInk** [comment]
I understand what you're saying and get why a window _was_ briefly appearing on a second monitor on a workspace switch. I say "was" because I just verified that's not happening on a two screen setup, with the screens side-by-side and moving workspaces left-to-right and vice-versa. It would appear that is _actually_ fixed as I semi-suspected (honestly a little surprised). The animation does only appear on the active monitor where the workspace is switching now, even if there are windows in the direction of the second monitor.

With that in mind, I think we're talking past each other a bit on this. I could still be mistaken, but it doesn't seem those two issues are related, especially given that I've confirmed the one you're talking about appears to be fixed (`Hyprland 0.46.2 commit 0bd541f2fd902dbfa04c3ea2ccf679395e316887`).

---

First, to establish **core functionality**: `gestures.workspace_swipe_fingers` and `scroller.gesture_scroll_enable` (with 4 fingers by default for workspaces) are doing the same **action** under scroller: using a touch-based swipe to change the workspace. The former uses a touchscreen and the latter uses the touchpad; otherwise they're _identical_ in function, but *not* behavior.

Second, in vanilla Hyprland: `gestures.workspace_swipe_fingers` and `gestures.workspace_swipe` are doing the same **action**: using a touch-based swipe to change the workspace. The former uses a touchscreen and the latter uses the touchpad just as before; otherwise they're _identical_ in function *and* behavior

Third, to establish what the desired **behavior** is (which is the problem I'm talking about): While the user is actively gesturing on the touch device, it is as if they are physically moving the workspace with their fingers. It moves directly *with* their fingers, as long as their fingers remain on the touch device. Once the user removes their fingers, the **action** is triggered, which sets off the final animation as defined as set up in the config, and it sets the workspace.

So:
- All three have the same **action**: they change the workspace via swiping on a touch device. This works as expected.
- Both `gestures.workspace_swipe_fingers` and `gestures.workspace_swipe` have the same **behavior** under Dwindle/Master. The workspace moves with the users fingers until they remove them, ending the interaction. This works as expected.
- `gestures.workspace_swipe_fingers` **has the same behavior while _scroller_ is active with no bugs or issues**. It works *exactly* the same as it does under Dwindle and Master. This works as expected
- `scroller.gesture_scroll_enabled` *does not* have the same behavior as `gestures.workspace_swipe_fingers` does, despite the fact that the **behavior** in question works 100% the same under scroller as it does Dwindle and Master when `gestures.workspace_swipe_fingers` is being used.

Putting that into a table, maybe it'll visually make a little more sense...

(blank cell indicates it _cannot_ be used in that environment or is not available)
| Config Name | Dwindle/Master | Scroller |
| --- | --- | --- |
| gestures.workspace_swipe_fingers | YES | YES |
| gestures.workspace_swipe | YES | |
| scroller.gesture_scroll_enabled (4 fingers) | | NO |

So, flowing from the fact that these three settings have the same **action**, ideally they should all three have the same **behavior**. It is _expected_ that `gestures.workspace_swipe` doesn't work at all in Scroller as it is disabled per a commit you made a while... What is unexpected in my view is that `gesture.workspace_swipe_fingers` under Scroller is capable of having the same **behavior** as `gesture.workspace_swipe`, but `scroller.gesture_scroll_enabled` is incapable of having that very same behavior when they are performing the same action.

In short: I would strongly assume that since `gestures.workspace_swipe_fingers` has no problem following the user's fingers (in the scroller layout) while the user is interacting with the touch device that `scroller.gesture_scroll_enabled` should be able to do the same thing. It doesn't appear to be an animation limitation of scroller since it does work via the touch screen gesture.

This, I would anticipate that table above to look like this:

| Config Name | Dwindle/Master | Scroller |
| --- | --- | --- |
| gestures.workspace_swipe_fingers | YES | YES |
| gestures.workspace_swipe | YES | |
| scroller.gesture_scroll_enabled (4 fingers) | | YES |

Hopefully that makes sense...I'm not sure how else to explain it. If you're still not getting what I'm getting at I can try and look at the code this evening, but I have no idea whether or not I'll find the missing link since I typically don't work with C++.

Here's the relevant config files, `hyprctl version` output, and `hyprpm list` output for good measure.

```ini
# 05-hypr-input.conf
cursor {
    use_cpu_buffer = true
}

gestures {
    ### Hyprland's touchpad workspace swipe is disabled and irrelevant ###
    workspace_swipe = false
    workspace_swipe_fingers = 3
    ### *This* is functioning as expected while **scroller** is active, demonstrated in video ###
    workspace_swipe_touch = true
}

input {
    accel_profile = adaptive
    sensitivity = 0

    touchpad {
        natural_scroll = true
        disable_while_typing = true
        middle_button_emulation = true
        clickfinger_behavior = true
        tap-to-click = true
    }
}
```
```
# 05-layout-scroller.conf
plugin {
    scroller {
        column_default_width = twothirds
        row_default_width = one
        ### This is enabled and I suspect it should behave like workspace_swipe_touch ###
        gesture_scroll_enable = true
        gesture_scroll_fingers = 3
        gesture_scroll_distance = 60
        gesture_overview_enable = true
        gesture_overview_fingers = 4
        gesture_overview_distance = 5
        center_row_if_space_available = true
    }
}
```
```
→ Repository hy3:
  │ Plugin hy3
  └─ enabled: false

→ Repository hyprland-plugins:
  │ Plugin borders-plus-plus
  └─ enabled: false

  │ Plugin csgo-vulkan-fix
  └─ enabled: false

  │ Plugin hyprbars
  └─ enabled: false

  │ Plugin hyprexpo
  └─ enabled: false

  │ Plugin hyprtrails
  └─ enabled: false

  │ Plugin hyprwinwrap
  └─ enabled: false

  │ Plugin xtra-dispatchers
  └─ enabled: false

→ Repository hyprscroller:
  │ Plugin hyprscroller
  └─ enabled: true
```
```
Hyprland 0.46.2 built from branch  at commit 0bd541f2fd902dbfa04c3ea2ccf679395e316887  (version: bump to 0.46.2).
Date: Thu Dec 19 19:26:47 2024
Tag: v0.46.2, commits: 5566
built against:
 aquamarine 0.7.1
 hyprlang 0.6.0
 hyprutils 0.3.3
 hyprcursor 0.1.11
 hyprgraphics 0.1.1


flags set:
debug
```

**Comment by dawsers** [comment]
I understand where you say `fingers` you are referring to `touch`? `fingers` is just the number of fingers to use.

I cannot test a touch screen, but I will try to see what they are doing.

In any case, they may have fixed something I am not aware of, like you said. So I will install the touchpad again and let you know what I find out.

**Comment by BinaryInk** [comment]
I _think_ I found what needs to be incorporated, just a matter of putting it together.
